Rubberduck.CodeAnalysis/Inspections/Concrete/MisleadingByRefParameterInspection.cs

```diff
@@ -0,0 +1,72 @@
+﻿using Rubberduck.CodeAnalysis.Inspections.Abstract;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.Parsing.VBA.DeclarationCaching;
+using Rubberduck.Resources.Inspections;
+using System.Linq;
+
+namespace Rubberduck.CodeAnalysis.Inspections.Concrete
+{
+    /// <summary>
+    /// Flags the value-parameter of a property mutators that are declared with an explict ByRef modifier.
+    /// </summary>
+    /// <why>
+    /// Regardless of the presence or absence of an explicit ByRef or ByVal modifier, the value-parameter
+    /// of a property mutator is always treated as though it had an explicit ByVal modifier.
+    /// Exception: UserDefinedType parameters are always passed by reference.
+    /// </why>
+    /// <example hasResult="true">
+    /// <module name="MyModule" type="Standard Module">
+    /// <![CDATA[
+    /// Private fizzField As Long
+    /// Public Property Get Fizz() As Long
+    ///     Fizz = fizzFiled
+    /// End Property
+    /// Public Property Let Fizz(ByRef arg As Long)
+    ///     fizzFiled = arg
+    /// End Property
+    /// ]]>
+    /// </module>
+    /// </example>
+    /// <example hasResult="false">
+    /// <module name="MyModule" type="Standard Module">
+    /// <![CDATA[
+    /// Private fizzField As Long
+    /// Public Property Get Fizz() As Long
+    ///     Fizz = fizzFiled
+    /// End Property
+    /// Public Property Let Fizz(arg As Long)
+    ///     fizzFiled = arg
+    /// End Property
+    /// ]]>
+    /// </module>
+    /// </example>
+    internal sealed class MisleadingByRefParameterInspection : DeclarationInspectionBase
+    {
+        public MisleadingByRefParameterInspection(IDeclarationFinderProvider declarationFinderProvider)
+            : base(declarationFinderProvider, DeclarationType.Parameter)
+        { }
+
+        protected override bool IsResultDeclaration(Declaration declaration, DeclarationFinder finder)
+        {
+            if ((declaration is ParameterDeclaration parameter)  
+                && parameter.ParentDeclaration is ModuleBodyElementDeclaration enclosingMethod
+                && (enclosingMethod.DeclarationType.HasFlag(DeclarationType.PropertyLet)
+                    || enclosingMethod.DeclarationType.HasFlag(DeclarationType.PropertySet))
+                && enclosingMethod.Parameters.Last() == parameter
+                && !(parameter.AsTypeDeclaration?.DeclarationType.HasFlag(DeclarationType.UserDefinedType) ?? false))
+            {
```

Comment:
Three remarks:
1. This if statement and the two returns can be combined into a single return statement. 
2. Can we please put the check for the flags on the parameter to before the type check on the parent? These checks should be comparatively fast.
3. I wonder whether we should (in a future PR) add a member `IsValueParameter` to `ParameterDeclaration` and set it in the declaration resolver. That should be possible to do efficiently there and would save us from getting the last parameter over and over.

Reply:
1. Yes
2. Yes
3. The `UserDefinedType` exception makes the `ValueParameter` a `ByRef` parameter...so, returning either `true` or `false` for an `IsValueParameter` member for that scenario seems incorrect.  Perhaps just adding a `IsLastParameter` to `ParameterDeclaration` achieves the improved efficiency and avoids passing judgement on the parameter mechanism.

Reply:
I think my naming for the property was a bit off. What I meant was `IsRhsParameter`, i.e. the last parameter of a mutator property.